project/JavalibIRCleaner.scala

```diff
@@ -0,0 +1,404 @@
+        validateClassName(interface.name)
+
+      Hashers.hashClassDef(this.transformClassDef(preprocessedTree))
+    }
+
+    private def transformInterfaceList(interfaces: List[Ident]): List[Ident] = {
+      /* Replace references to scala.Serializable by java.io.Serializable.
+       * This works around the fact that scalac adds scala.Serializable to the
+       * companion object of any class that extends java.io.Serializable.
+       */
+
+      val ScalaSerializable = "s_Serializable"
+      val JavaIOSerializable = "Ljava_io_Serializable"
+
+      if (!interfaces.exists(_.name == ScalaSerializable)) {
+        interfaces
+      } else if (interfaces.exists(_.name == JavaIOSerializable)) {
+        interfaces.filter(_.name != "s_Serializable")
+      } else {
+        interfaces.map { ident =>
+          if (ident.name == ScalaSerializable)
+            Ident(JavaIOSerializable, Some("java.io.Serializable"))(ident.pos)
+          else
+            ident
+        }
+      }
+    }
+
+    override def transformMemberDef(memberDef: MemberDef): MemberDef = {
+      super.transformMemberDef(memberDef) match {
+        case m: MethodDef =>
+          m.name match {
+            case name: Ident =>
+              MethodDef(m.flags, transformMethodIdent(name), m.args,
+                  m.resultType, m.body)(m.optimizerHints, m.hash)(m.pos)
+            case _ =>
+              m
+          }
+        case m =>
+          m
+      }
+    }
+
+    override def transform(tree: Tree, isStat: Boolean): Tree = {
+      implicit val pos = tree.pos
+
+      val preprocessedTree = tree match {
+        /* Rewrite `throw sjs.runtime.unwrapJavaScriptExeption(arg)` to
+         * `throw arg`, which is OK because the Java lang lib never throws
+         * wrapped JavaScript exceptions.
+         */
+        case Throw(Apply(_, LoadModule(ClassRef("sjsr_package$")),
+            Ident("unwrapJavaScriptException__jl_Throwable__O", _), arg :: Nil)) =>
+          Throw(arg)
+
+        case _ =>
+          tree
+      }
+
+      val result = super.transform(preprocessedTree, isStat) match {
+        case New(cls, ctor, args) =>
+          New(cls, transformMethodIdent(ctor), args)
+
+        case t: Apply =>
+          Apply(t.flags, t.receiver, transformMethodIdent(t.method),
+              t.args)(t.tpe)
+        case t:ApplyStatically =>
+          ApplyStatically(t.flags, t.receiver, transformClassRef(t.cls),
+              transformMethodIdent(t.method), t.args)(t.tpe)
+        case t: ApplyStatic =>
+          ApplyStatic(t.flags, transformClassRef(t.cls),
+              transformMethodIdent(t.method), t.args)(t.tpe)
+
+        case NewArray(typeRef, lengths) =>
+          NewArray(transformArrayTypeRef(typeRef), lengths)
+        case ArrayValue(typeRef, elems) =>
+          ArrayValue(transformArrayTypeRef(typeRef), elems)
+
+        case IsInstanceOf(expr, typeRef) =>
+          IsInstanceOf(expr, transformTypeRef(typeRef))
```

Comment:
This is also bad: It breaks the semantics of `isInstnaceOf`. We should fail if we get a JS class here.